### PR TITLE
Linux: Cross-platform user namespace xattrs compat

### PR DESCRIPTION
**Pull request by freqlabs**
ZFS on Linux originally implemented xattr namespaces in a way that is
incompatible with other operating systems.  On illumos, xattrs do not
have namespaces.  Every xattr name is visible.  FreeBSD has two
universally defined namespaces: EXTATTR_NAMESPACE_USER and
EXTATTR_NAMESPACE_SYSTEM.  The system namespace is used for protected
FreeBSD-specific attributes such as MAC labels and pnfs state.  These
attributes have the namespace string "freebsd:system:" prefixed to the
name in the encoding scheme used by ZFS.  The user namespace is used
for general purpose user attributes and obeys normal access control
mechanisms.  These attributes have no namespace string prefixed, so
xattrs written on illumos are accessible in the user namespace on
FreeBSD, and xattrs written to the user namespace on FreeBSD are
accessible by the same name on illumos.

Linux has several xattr namespaces.  For Linux, ZFS encodes the namespace
in the xattr name for every namespace, including the user namespace.  As a
consequence, an xattr in the user namespace with the name "foo" is stored
by ZFS with the name "user.foo" and therefore appears on FreeBSD and
illumos to have the name "user.foo" rather than "foo".  Conversely, none of the
xattrs written on FreeBSD or illumos are accessible on Linux unless the name
happens to be prefixed with one of the Linux xattr namespaces, in which
case the namespace is stripped from the name.  This makes xattrs
entirely incompatible between Linux and other platforms.

We want to make the encoding of user namespace xattrs compatible across
platforms.  A critical requirement of this compatibility is for xattrs
from existing pools from FreeBSD and illumos to be accessible by the
same names in the user namespace on Linux.  It is also necessary that
existing pools with xattrs written by Linux retain access to those
xattrs by the same names on Linux.  Making user namespace xattrs from
Linux accessible by the correct names on other platforms is important.
The handling of other namespaces is not required to be consistent.

Add a fallback mechanism for listing and getting xattrs to treat xattrs
as being in the user namespace if they do not match a known prefix.

When setting user namespace xattrs, do not prefix the namespace to the
name.  If the xattr is already present with the namespace prefix,
remove it so only the non-prefixed version persists.  This ensures
other platforms will be able to read the xattr with the correct name.

Explicitly ignore freebsd:system namespace xattrs.

TODO:

* [x] If xattrs with the user namespace prefix are already present, they
  will not be automatically fixed.  Any existing xattrs must be
  manually rewritten on Linux for the name to be correct on FreeBSD.
  This also means that files may have a mix of incompatible and
  compatible names.  Other platforms could strip the Linux user
  namespace prefix from xattr names so they are presented correctly.

* [x] The newly written xattrs will no longer be visible on previous
  versions of ZFS on Linux.  This behavior needs to be made optional
  with a feature flag and possibly a per-dataset property.

* [ ] There is no attempt to handle xattr names that clash with a namespace
  prefix.  If you write an xattr named "user.foo" to the user namespace
  on FreeBSD, the "user." prefix will be stripped on Linux.  This was
  partially the case already, except now the stripped name will also
  replace the prefixed name when updating the xattr.  Likewise, setting
  an xattr to the user namespace using a name with the prefix of
  another namespace may cause the xattr to be manipulated in the other
  namespace.  This is potentially a security issue.  Such names must be
  forbidden.

* [ ] New tests should be added when the functionality is complete.

* [x] Documentation will be needed.

### CODE REVIEW

**Comment by freqlabs** [comment]
* Rebased

**Review by rick-mesta — APPROVED** [review]
I was going to make a quip about the abi files having your $HOME directory embedded on them, but then the next commit fixed that right up 😁 . This LGTM 👍🏼

File: module/os/freebsd/zfs/zfs_vnops_os.c [hunk 1]
```diff
@@ -5872,18 +5982,16 @@ zfs_listextattr(struct vop_listextattr_args *ap)
...
-
 	ZFS_ENTER(zfsvfs);
 	ZFS_VERIFY_ZP(zp);
 	rw_enter(&zp->z_xattr_lock, RW_READER);
-	if (zfsvfs->z_use_sa && zp->z_is_sa)
-		error = zfs_listextattr_sa(ap, attrprefix);
-	if (error == 0)
-		error = zfs_listextattr_dir(ap, attrprefix);
+
+	boolean_t compat = !!(zfsvfs->z_flags & ZSB_XATTR_COMPAT);
```
[the remark this answers is not in the document]
Comment (freqlabs) [reply module/os/freebsd/zfs/zfs_vnops_os.c:5989]: It's a short hand way of converting the bit field to a boolean value, `!!(0b1010 & 0b1000)` -> `!!0b1000` -> `!0b0` -> `0b1`

[the remark this answers is not in the document]
Comment (rick-mesta) [reply module/os/freebsd/zfs/zfs_vnops_os.c:5989]: 👍🏼

File: include/sys/fs/zfs.h [hunk 2]
```diff
@@ -1613,6 +1619,36 @@ typedef enum {
...
 
+/*
+ * xattr namespace prefixes.  These are forbidden in xattr names.
+ *
+ * For cross-platform compatibility, xattrs in the user namespace should not be
+ * prefixed with the namespace name, but for backwards compatibility with older
+ * ZFS on Linux versions we do prefix the namespace.
+ */
+#define	ZFS_XA_NS_FREEBSD_PREFIX		"freebsd:"
+#define	ZFS_XA_NS_FREEBSD_PREFIX_LEN		strlen("freebsd:")
```
Comment (yocalebo) [inline include/sys/fs/zfs.h:1630]: Obviously these prefixes don't change but just curious on why not do `ZFS_XA_NS_FREEBSD_PREFIX_LEN strlen(ZFS_XA_NS_FREEBSD_PREFIX)`

Comment (freqlabs) [reply include/sys/fs/zfs.h:1630]: Just to avoid the line wrap.

**Comment by freqlabs** [comment]
I'm working on a few last fixups and squashing the commits, then if nobody objects I'll be ready to merge this.

File: module/os/freebsd/zfs/zfs_vnops_os.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (amotin) [reply module/os/freebsd/zfs/zfs_vnops_os.c:6006]: Some consider those as Linux'isms.  Discouraged in FreeBSD code, recommending != 0 instead.

**Pull request merged**